packages/beacon-state-transition/src/altair/block/processSyncCommittee.ts

```diff
@@ -1,59 +1,86 @@
 import {BitList, isTreeBacked, TreeBacked} from "@chainsafe/ssz";
 
 export function processSyncCommittee(
   state: CachedBeaconState<altair.BeaconState>,
-  aggregate: altair.SyncAggregate,
+  syncAggregate: altair.SyncAggregate,
   verifySignatures = true
 ): void {
-  const {config, epochCtx} = state;
-  const {syncParticipantReward, syncProposerReward} = epochCtx;
-  const previousSlot = Math.max(state.slot, 1) - 1;
-  const committeeIndices = state.currSyncCommitteeIndexes;
-  // the only time aggregate is not a TreeBacked is when producing a new block
-  const participantIndices = isTreeBacked(aggregate)
-    ? zipIndexesInBitList(
-        committeeIndices,
-        aggregate.syncCommitteeBits as TreeBacked<BitList>,
-        config.types.altair.SyncCommitteeBits
-      )
-    : committeeIndices.filter((index) => !!aggregate.syncCommitteeBits[index]);
-  const participantPubkeys = participantIndices.map((validatorIndex) => state.validators[validatorIndex].pubkey);
-  const epochSig = computeEpochAtSlot(config, previousSlot);
-  const domain = getDomain(config, state, config.params.DOMAIN_SYNC_COMMITTEE, epochSig);
-  const signingRoot = computeSigningRoot(
-    config,
-    config.types.Root,
-    getBlockRootAtSlot(config, state, previousSlot),
-    domain
-  );
+  const {syncParticipantReward, syncProposerReward} = state.epochCtx;
+  const participantIndices = getParticipantIndices(state, syncAggregate);
+
   // different from the spec but not sure how to get through signature verification for default/empty SyncAggregate in the spec test
-  if (verifySignatures && participantIndices.length > 0) {
-    assert.true(
-      verifyAggregate(
-        participantPubkeys.map((pubkey) => pubkey.valueOf() as Uint8Array),
-        signingRoot,
-        aggregate.syncCommitteeSignature.valueOf() as Uint8Array
-      ),
-      "Sync committee signature invalid"
-    );
+  if (verifySignatures) {
+    const signatureSet = getSyncCommitteeSignatureSet(state, syncAggregate, participantIndices);
+    // When there's no participation we consider the signature valid and just ignore i
+    if (signatureSet !== null && !verifySignatureSet(signatureSet)) {
+      throw Error("Sync committee signature invalid");
+    }
   }
 
-  const proposerIndex = epochCtx.getBeaconProposer(state.slot);
+  const proposerIndex = state.epochCtx.getBeaconProposer(state.slot);
   for (const participantIndex of participantIndices) {
     increaseBalance(state, participantIndex, syncParticipantReward);
   }
   increaseBalance(state, proposerIndex, syncProposerReward * BigInt(participantIndices.length));
 }
+
+export function getSyncCommitteeSignatureSet(
+  state: CachedBeaconState<altair.BeaconState>,
+  syncAggregate: altair.SyncAggregate,
+  /** Optional parameter to prevent computing it twice */
+  participantIndices?: number[]
+): ISignatureSet | null {
+  const {config, epochCtx} = state;
+
+  const previousSlot = Math.max(state.slot, 1) - 1;
```

Comment:
@dapplion previousSlot should be from `Math.max(block.slot, 1) - 1`

Reply:
Good pick!